src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/DependencyGraphRestoreUtility.cs

```diff
@@ -0,0 +1,111 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using NuGet.Commands;
+using NuGet.Common;
+using NuGet.Configuration;
+using NuGet.Packaging;
+using NuGet.Packaging.Core;
+using NuGet.ProjectManagement;
+using NuGet.ProjectModel;
+using NuGet.Protocol;
+using NuGet.Protocol.Core.Types;
+
+namespace NuGet.PackageManagement
+{
+    public static class DependencyGraphRestoreUtility
+    {
+        public static async Task<bool> IsRestoreRequiredAsync(
+            IEnumerable<IDependencyGraphProject> projects,
+            bool forceRestore,
+            INuGetPathContext pathContext,
+            ExternalProjectReferenceContext referenceContext)
+        {
+            // Swap caches
+            var oldProjectCache = referenceContext.ProjectCache;
+            referenceContext.ProjectCache = await DependencyGraphProjectCacheUtility.CreateDependencyGraphProjectCache(
+                projects,
+                referenceContext);
+
+            if (forceRestore)
+            {
+                // The cache has been updated, now skip the check since we are doing a restore anyways.
+                return true;
+            }
+
+            if (DependencyGraphProjectCacheUtility.CacheHasChanges(oldProjectCache, referenceContext.ProjectCache))
+            {
+                // A new project has been added
+                return true;
+            }
+            
+            // Read package folder locations, initializing them in order of priority
+            var packageFolderPaths = new List<string>();
+            packageFolderPaths.Add(pathContext.UserPackageFolder);
+            packageFolderPaths.AddRange(pathContext.FallbackPackageFolders);
+            var pathResolvers = packageFolderPaths.Select(path => new VersionFolderPathResolver(path));
+
+            var packagesChecked = new HashSet<PackageIdentity>();
+            if (projects.Any(p => p.IsRestoreRequired(pathResolvers, packagesChecked, referenceContext)))
```

Comment:
So that it is shared across multiple invocations of `IsRestoreRequired`. If a package is references by more than one of the `projects`, there is no need to check if that package is on disk more than once.